rllib/contrib/bandits/agents/linear.py

```diff
@@ -0,0 +1,164 @@
+import copy
+import time
+from enum import Enum
+
+from gym import spaces
+from ray.rllib.agents.trainer import with_common_config
+from ray.rllib.agents.trainer_template import build_trainer
+from ray.rllib.contrib.bandits.exploration import ThompsonSampling, UCB
+from ray.rllib.contrib.bandits.models.linear_regression import \
+    DiscreteLinearModelThompsonSampling, \
+    DiscreteLinearModelUCB, DiscreteLinearModel, \
+    ParametricLinearModelThompsonSampling, ParametricLinearModelUCB
+from ray.rllib.models.catalog import ModelCatalog
+from ray.rllib.models.model import restore_original_dimensions
+from ray.rllib.policy.policy import Policy, LEARNER_STATS_KEY
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.policy.torch_policy import TorchPolicy
+from ray.rllib.policy.torch_policy_template import build_torch_policy
+from ray.rllib.utils.annotations import override
+
+# yapf: disable
+# __sphinx_doc_begin__
+DEFAULT_CONFIG = with_common_config({
+    # No remote workers by default.
+    "num_workers": 0,
+    "use_pytorch": True,
+
+    # Do online learning
+    "sample_batch_size": 1,
+    "train_batch_size": 1,
+
+    # Bandits cant afford to do one timestep per iteration as it is extremely
+    # slow because of metrics collection overhead. This setting means that the
+    # agent will be trained for 100 times in one iteration of Rllib
+    "timesteps_per_iteration": 100
+})
+
+
+# __sphinx_doc_end__
+# yapf: enable
+
+
+class ImplicitExploration(Enum):
+    ThompsonSampling = 1
+    UCB = 2
+
+
+class BanditPolicyOverrides:
+    @override(Policy)
+    def _create_exploration(self, action_space, config):
+        exploration_config = config.get("exploration_config",
+                                        {"type": "StochasticSampling"})
+        if exploration_config[
+                "type"] == ImplicitExploration.ThompsonSampling.name:
```

Comment:
@sven1977 is this the right way to create ThompsonSampling or it something that can be auto-created from_config?

Reply:
Yes, we should create it from the config. I guess the problem is that `from_config` only looks in the package in which the base Exploration class is located and thus would miss these new custom classes. However, you can always do:
```
exploration_config: {
"type": "ray.rllib.contrib.bandits.exploration. ImplicitExploration.ThompsonSampling"
}
```

Reply:
@sven1977 could you push the patch for this?

Reply:
I just pushed the patch for this,